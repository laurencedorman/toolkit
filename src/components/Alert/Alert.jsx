/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Transition, animated, config } from 'react-spring';
import cn from 'classnames';

import Hint from '../Hint';
import Portal from '../Portal';
import styles from './Alert.module.scss';

export const translate = value => {
  switch (value) {
    case 'topRight':
    case 'bottomRight':
      return 320;
    case 'topLeft':
    case 'bottomLeft':
      return -320;
    default:
      return 320;
  }
};

const timer = (toggle, timeout) => setTimeout(() => toggle(), timeout);

const Alert = ({
  on,
  className,
  toggle,
  theme,
  textAlert,
  timeout,
  position,
  dataQa,
}) => {
  const classNames = cn(
    'toolkit',
    styles.alert,
    {
      [styles.right]: position === 'topRight' || position === 'bottomRight',
      [styles.left]: position === 'topLeft' || position === 'bottomLeft',
      [styles.top]: position === 'topRight' || position === 'topLeft',
      [styles.bottom]: position === 'bottomRight' || position === 'bottomLeft',
    },
    className
  );

  const setTranslate = translate(position);

  if (timeout !== undefined && on) {
    timer(toggle, timeout);
  }

  return (
    <Portal>
      <Transition
        native
        items={on}
        config={config.stiff}
        from={{ transform: `translateX(${setTranslate}px)` }}
        enter={{ transform: 'translateX(0)' }}
        leave={{ transform: `translateX(${setTranslate}px)` }}
      >
        {display =>
          display &&
          // eslint-disable-next-line react/display-name
          (style => (
            <animated.div
              className={classNames}
              role="contentinfo"
              style={style}
              onClick={toggle}
              timer={timeout}
              position={position}
            >
              <Hint textAlert={textAlert} theme={theme} dataQa={dataQa} />
            </animated.div>
          ))
        }
      </Transition>
    </Portal>
  );
};

Alert.displayName = 'Alert';

Alert.propTypes = {
  /** display alert */
  on: PropTypes.bool.isRequired,
  /** Required function to set props.on (display) */
  toggle: PropTypes.func.isRequired,
  theme: PropTypes.oneOf(['default', 'light', 'danger', 'menthe']),
  textAlert: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  className: PropTypes.string,
  /** Set timer in ms to remove automatically alert */
  timeout: PropTypes.number,
  position: PropTypes.oneOf([
    'topLeft',
    'bottomLeft',
    'topRight',
    'bottomRight',
  ]),
  dataQa: PropTypes.string,
};

Alert.defaultProps = {
  theme: 'default',
  className: '',
  position: 'topRight',
  timeout: 0,
  dataQa: '',
};

export default Alert;
