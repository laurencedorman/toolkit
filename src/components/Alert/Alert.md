Warning

- It includes react-spring as dependency (31.79KB)

Alert:

- Display an alert with Hint Component.
- You can use Toggle component to easy display. Read <b>Toggle</b> section to have more information.
- Alert use <b>Portal</b> component who needs some adding to work. Read Portal section to get more information.
- Set different theme `[ 'default' | 'light' | 'danger' | 'menthe' ]`.
- Change icon.
- Set timeout in ms to remove automatically alert.
- Set position `[ 'topRight' | 'bottomRight' | 'topLeft' | 'bottomLeft' ]`.

Tips:

- Force toggleOut on toggle props to avoid conflict between close click event and close timeout function

```js
import Button from '../Button';
import Toggle from '../Toggle';

<Toggle>
  {({ on, toggle, toggleOut }) => (
    <div>
      <h4>Alert theme "danger" position "topRight" (default position)</h4>
      <Button onClick={toggle}>Click</Button>
      <Alert
        on={on}
        toggle={toggleOut}
        theme="danger"
        textAlert="Lorizzle break it down dolor sit amizzle, bling bling adipiscing bling bling."
        timeout={3000}
      />
    </div>
  )}
</Toggle>;
```

```js
import Button from '../Button';

initialState = {
  on: false,
};

<div>
  <h4>
    Alert theme "default" position "bottomRight" without using Toggle component
  </h4>
  <Button onClick={() => setState({ on: true })}>Click</Button>
  <Alert
    on={state.on}
    toggle={() => setState({ on: false })}
    textAlert="Lorizzle break it down dolor sit amizzle, bling bling adipiscing bling bling."
    timeout={3000}
    position="bottomRight"
  />
</div>;
```

```js
import Button from '../Button';
import Toggle from '../Toggle';

<Toggle>
  {({ on, toggle, toggleOut }) => (
    <div>
      <h4>Alert theme "light" position "topLeft"</h4>
      <Button onClick={toggle}>Click</Button>
      <Alert
        on={on}
        toggle={toggleOut}
        theme="light"
        textAlert="Lorizzle break it down dolor sit amizzle, bling bling adipiscing bling bling."
        timeout={3000}
        position="topLeft"
      />
    </div>
  )}
</Toggle>;
```

```js
import Button from '../Button';
import Toggle from '../Toggle';

<Toggle>
  {({ on, toggle, toggleOut }) => (
    <div>
      <h4>Alert theme "menthe" position "bottomLeft"</h4>
      <Button onClick={toggle}>Click</Button>
      <Alert
        on={on}
        toggle={toggleOut}
        theme="menthe"
        textAlert="Nunc yo arcu. Donec ante. We gonna chung placerizzle mauris at lectus. Morbi fo shizzle tortizzle. Crizzle yo mamma tortor nizzle mi consectetizzle sure. Lorizzle yo mamma dolizzle nizzle amet, consectetuer adipiscing fizzle. Ma nizzle fizzle arcu, things uhuh ... yih!, eleifend izzle, mollizzle in, est. Aliquizzle erat volutpizzle."
        timeout={3000}
        position="bottomLeft"
      />
    </div>
  )}
</Toggle>;
```
