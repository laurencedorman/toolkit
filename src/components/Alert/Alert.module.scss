/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import '../../styles/import.scss';

.alert {
  position: fixed;
  width: em(300);
  cursor: pointer;
  z-index: z($layers, modal);

  &.right {
    right: em($space);
  }
  &.left {
    left: em($space);
  }
  &.top {
    top: em($space);
  }
  &.bottom {
    bottom: em($space);
  }
}
