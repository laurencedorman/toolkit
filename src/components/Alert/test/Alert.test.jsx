/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Alert, { translate } from '../Alert';

describe('Alert', () => {
  let wrapper;
  const toggle = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Alert on={false} toggle={toggle} textAlert="test" theme="danger" />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render without crash with layoutContainer', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('should render alert if props on is true', () => {
    wrapper = mount(<Alert on textAlert="test" toggle={toggle} />);
    expect(wrapper.props().on).toBe(true);
  });

  it('should render timeout close function if props timeout is declared', () => {
    wrapper = mount(
      <Alert on timeout={300} textAlert="test" toggle={toggle} />
    );
    expect(wrapper.props().timeout).toEqual(300);
  });

  it('should set position on bottom right if position is changed', () => {
    wrapper = mount(
      <Alert on position="bottomRight" textAlert="test" toggle={toggle} />
    );
    expect(wrapper.props().position).toEqual('bottomRight');
  });

  it('should set position on top left if position is changed', () => {
    wrapper = mount(
      <Alert on position="topLeft" textAlert="test" toggle={toggle} />
    );
    expect(wrapper.props().position).toEqual('topLeft');
  });

  it('should set position on bottom left if position is changed', () => {
    wrapper = mount(
      <Alert on position="bottomLeft" textAlert="test" toggle={toggle} />
    );
    expect(wrapper.props().position).toEqual('bottomLeft');
  });

  it('should display with toggle', () => {
    wrapper = mount(<Alert on textAlert="test" toggle={toggle} />);
    expect(wrapper.props().toggle).toEqual(toggle);
  });

  test('translate', () => {
    expect(translate('default')).toBe(320);
  });
});
