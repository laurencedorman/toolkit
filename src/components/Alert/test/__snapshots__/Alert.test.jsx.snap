// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should match snapshot 1`] = `
<Portal>
  <Transition
    config={
      Object {
        "friction": 20,
        "tension": 210,
      }
    }
    enter={
      Object {
        "transform": "translateX(0)",
      }
    }
    from={
      Object {
        "transform": "translateX(320px)",
      }
    }
    items={false}
    keys={[Function]}
    leave={
      Object {
        "transform": "translateX(320px)",
      }
    }
    native={true}
    reset={false}
    unique={false}
  >
    <Component />
  </Transition>
</Portal>
`;
